modules/frontend/frontend.go

```diff
@@ -97,76 +180,62 @@ func NewTripperware(cfg Config, logger log.Logger, registerer prometheus.Registe
 				}
 				resp.Body = ioutil.NopCloser(bytes.NewReader(jsonTrace.Bytes()))
 			}
-
 			span.SetTag("response marshalling format", marshallingFormat)
 
-			traceID, _ := middleware.ExtractTraceID(ctx)
-			statusCode := 500
-			var contentLength int64 = 0
-			if resp != nil {
-				statusCode = resp.StatusCode
-				contentLength = resp.ContentLength
-			}
-
-			level.Info(logger).Log(
-				"tenant", orgID,
-				"method", r.Method,
-				"traceID", traceID,
-				"url", r.URL.RequestURI(),
-				"duration", time.Since(start).String(),
-				"response_size", contentLength,
-				"status", statusCode,
-			)
-
 			return resp, err
 		})
-	}, nil
+	}
 }
 
-type Handler interface {
-	Do(*http.Request) (*http.Response, error)
-}
+// NewSearchTripperware creates a new frontend tripperware to handle search and search tags requests.
+func NewSearchTripperware(logger log.Logger) queryrange.Tripperware {
+	return func(rt http.RoundTripper) http.RoundTripper {
+		return queryrange.RoundTripFunc(func(r *http.Request) (*http.Response, error) {
+			// don't start a new span, this is already handled by frontendRoundTripper
+			span := opentracing.SpanFromContext(r.Context())
 
-type Middleware interface {
-	Wrap(Handler) Handler
-}
+			// check marshalling format
+			marshallingFormat := util.JSONTypeHeaderValue
+			if r.Header.Get(util.AcceptHeaderKey) == util.ProtobufTypeHeaderValue {
+				marshallingFormat = util.ProtobufTypeHeaderValue
+			}
 
-// MiddlewareFunc is like http.HandlerFunc, but for Middleware.
-type MiddlewareFunc func(Handler) Handler
+			// TODO GRPC requests aren't handled properply yet
```

Comment:
Yeah, I'm kind of annoyed at this tbh. Having support for json is nice to view stuff in your browser/curl, but if we want to support json while using protobuf internally we would have to convert between them here. That's the commented out code underneath `rt.RoundTrip(r)`.

The difference between `NewTracesTripperware` and `NewSearchTripperware` is that search can return multiple data structs, so we would have to check what path we are on to determine which struct to unmarshal the data into.

Reply:
Personally I'd like to keep support for json in the query-frontend, just because it's a good for debugging issues when Grafana isn't working.

Reply:
There's also the option of adding search support to the CLI (that way we could only support protobufs)

But agree that having json is nice

> The difference between NewTracesTripperware and NewSearchTripperware is that search can return multiple data structs, so we would have to check what path we are on to determine which struct to unmarshal the data into.

Ah I see. In that case it might even be better to support only json :)

Reply:
For the traces path we only moved to proto because of the marshalling/unmarshalling costs. But those might be small in the search path.

Reply:
Only json would be the path of least resistance, it seems like a good tradeoff to leave it as is for now and add protobuf support later on.

Reply:
👍  agree. Let's drop the TODO in the comment. We can revisit it later if we see performance issues.